linera-web/src/signer.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module contains various implementation of the [`Signer`] trait usable in the browser.
+use std::{fmt::Display, str::FromStr};
+
+use linera_base::{
+    crypto::{AccountPublicKey, AccountSignature, CryptoHash, EvmPublicKey, EvmSignature, Signer},
+    identifiers::AccountOwner,
+};
+use wasm_bindgen::prelude::*;
+use web_sys::wasm_bindgen;
+
+#[repr(u8)]
+#[wasm_bindgen(js_name = "SignerError")]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum JsSignerError {
+    MissingKey = 0,
+    SigningError = 1,
+    PublicKeyParse = 2,
+    JsConversion = 3,
+    UnexpectedSignatureFormat = 4,
+    Unknown = 9,
+}
+
+impl Display for JsSignerError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            JsSignerError::MissingKey => write!(f, "No key found for the given owner"),
+            JsSignerError::SigningError => write!(f, "Error signing the value"),
+            JsSignerError::PublicKeyParse => write!(f, "Error parsing the public key"),
+            JsSignerError::JsConversion => write!(f, "Error converting JS value"),
+            JsSignerError::UnexpectedSignatureFormat => {
+                write!(f, "Unexpected signature format received from JS")
+            }
+            JsSignerError::Unknown => write!(f, "An unknown error occurred"),
+        }
+    }
+}
+
+impl From<JsValue> for JsSignerError {
+    fn from(value: JsValue) -> Self {
+        if let Some(fnum) = value.as_f64() {
+            match fnum as u8 {
+                0 => JsSignerError::MissingKey,
+                1 => JsSignerError::SigningError,
+                2 => JsSignerError::PublicKeyParse,
+                3 => JsSignerError::JsConversion,
+                4 => JsSignerError::UnexpectedSignatureFormat,
+                _ => JsSignerError::Unknown,
+            }
+        } else {
+            JsSignerError::Unknown
+        }
+    }
+}
+
+impl std::error::Error for JsSignerError {}
+
+// An interface that will be compiled to TypeScript and exported for use in the browser.
+#[wasm_bindgen(typescript_custom_section)]
+const SIGNER_INTERFACE: &'static str = include_str!("signer.d.ts");
+
+#[wasm_bindgen]
+extern "C" {
+    // We refer to the interface defined above.
+    #[wasm_bindgen(typescript_type = "Signer")]
+    pub type JsSigner;
+
+    #[wasm_bindgen(catch, method)]
+    async fn sign(this: &JsSigner, owner: JsValue, value: Vec<u8>) -> Result<JsValue, JsValue>;
+
+    #[wasm_bindgen(catch, method)]
+    async fn get_public_key(this: &JsSigner, owner: JsValue) -> Result<JsValue, JsValue>;
```

Comment:
Does this not require a rename (`js_name = getPublicKey`) to match the interface? (and below)

Reply:
Right, but if the user passes something like `{ getPublicKey: () => … }` and then we call `get_public_key` on it (which is what this code does) we'll get an error (‘`get_public_key` is undefined’).  `js_name = getPublicKey` means that when we call `get_public_key` on the Rust side we'll call the JavaScript function `getPublicKey`, which is what our TypeScript interface claims should exist.